src/feedback_forensics/app/model_annotators.py

```diff
@@ -0,0 +1,120 @@
+        reference_models: List of model names to use as reference models. Empty list means all models are used as references.
+
+    Returns:
+        Tuple of (annotator_metadata, df_with_annotators)
+    """
+    start_time = time.time()
+
+    all_targets = len(target_models) == 0
+    one_vs_all = len(reference_models) == 0
+    if all_targets or one_vs_all:
+        all_models = get_available_models(df)
+        if all_targets:
+            target_models = all_models
+        if one_vs_all:
+            reference_models = all_models
+
+    annotations_df = df[["comparison_id"]].copy()
+
+    if len(target_models) == 0:
+        logger.info(
+            "No target models available, creating no model identity annotators."
+        )
+        return {}, annotations_df
+
+    if len(reference_models) == 0:
+        logger.info("No reference models specified. Using all models as references.")
+
+    # Filter out rows with missing model data or same model on both sides
+    valid_mask = ~(
+        pd.isna(df["model_a"])
+        | pd.isna(df["model_b"])
+        | (df["model_a"] == df["model_b"])
+    )
+    valid_model_a = df.loc[valid_mask, "model_a"]
+    valid_model_b = df.loc[valid_mask, "model_b"]
+    valid_idx = df.index[valid_mask]
+
+    model_a_in_refs = valid_model_a.isin(reference_models)
+    model_b_in_refs = valid_model_b.isin(reference_models)
+
+    annotator_start = time.time()
+    prep_time = annotator_start - start_time
+    logger.debug(f"Model annotator preparation time: {prep_time:.2f} seconds")
+
+    # For each target model, create an annotator that always prefers that model over reference models
+    total_annotations = 0
+    annotator_metadata = {}
+
+    for model in target_models:
+        other_refs = [m for m in reference_models if m != model]
+        if len(other_refs) == 0:
+            continue
+
+        annotator_name = f"model_identity_{model}_over_references"
+        annotator_id = hash_string(annotator_name)
+
+        if one_vs_all:
+            description = f"Always prefer {model} over any other model"
+        else:
+            description = f"Always prefer {model} over reference models ({iter_to_trunc_str(other_refs, 3)})"
+
+        annotator_metadata[annotator_id] = {
+            "variant": MODEL_IDENTITY_ANNOTATOR_TYPE,
+            "model_id": model,
+            "reference_models": reference_models,
+            "annotator_visible_name": f"{PREFIX_MODEL_IDENTITY_ANNOTATORS}{model}",
+            "annotator_in_row_name": f"{model}-preference",
+            "annotator_description": description,
+        }
+
+        values = pd.Series("Not applicable", index=df.index)
+
+        # Create masks for when this model should be preferred (in column A or B)
+        a_preference_mask = (valid_model_a == model) & model_b_in_refs
+        b_preference_mask = (valid_model_b == model) & model_a_in_refs
```

Comment:
model_b_in_refs here could theoretically also include same model comparisons, e.g. Llama 3 vs Llama 3. I think we want to filter such same model comparisons? Could be fixed maybe by checking for and removing overlaps between a_preference_mask and b_preference mask?

Reply:
That's already filtered with `valid_mask`.